lib/puppet/application/device.rb

```diff
@@ -219,20 +238,36 @@ def main
           Puppet[:vardir] = ::File.join(Puppet[:devicedir], device.name)
           Puppet[:certname] = device.name
 
-          # this will reload and recompute default settings and create the devices sub vardir, or we hope so :-)
-          Puppet.settings.use :main, :agent, :ssl
-
           # this init the device singleton, so that the facts terminus
           # and the various network_device provider can use it
           Puppet::Util::NetworkDevice.init(device)
 
-          # ask for a ssl cert if needed, but at least
-          # setup the ssl system for this device.
-          setup_host
-
-          require 'puppet/configurer'
-          configurer = Puppet::Configurer.new
-          configurer.run(:network_device => true, :pluginsync => Puppet::Configurer.should_pluginsync?)
+          if options[:apply]
+            # avoid reporting to server
+            Puppet::Transaction::Report.indirection.terminus_class = :yaml
+            Puppet::Resource::Catalog.indirection.cache_class = nil
+
+            require 'puppet/application/apply'
+            begin
+
+              Puppet[:node_terminus] = :plain
+              Puppet[:catalog_terminus] = :compiler
+              Puppet[:catalog_cache_terminus] = nil
+              Puppet[:facts_terminus] = :network_device
```

Comment:
We're setting terminii and cache classes directly, e.g. `terminus_class = :yaml` and indirectly through settings `Puppet[:node_terminus] = :plain`. Can that be simplified to one or the other, e.g. we setting the catalog cache terminus in two places.

What side-effects will happen if you run `puppet device` and later run `puppet agent` (or the daemon kicks in)? For example, we're not setting the report cache terminus, so will puppet device save a report locally that "conflicts" with the local agent? Same for node cache terminus?

Reply:
@joshcooper i didnt initially want to change what was happening https://github.com/tphoney/puppet/blob/e1b856ea557e248c6f91fbc108f0620257edb749/lib/puppet/application/device.rb#L10-L16 we could have logic further down, once we have options to set them. though i am unsure of the ramifications.

"What side-effects will happen if you run puppet device and later run puppet agent" are you talking about if agent and device run at the same time, and the shared settings from device conflicts with agent ? From what i have seen each executable picks up its own defaults, unless you instantiate a puppet application from within another. IE what we are trying to do here. IE it takes a lot to mess with the default running of a single puppet application.

Reply:
It may be good to have a call tomorrow to talk through this ?

Reply:
I dug through this on Puppet 4 and 5.  If we have already setup the cache directory for the target everything functions as one would expect.  There does not appear to be any artifacts of cached catalog or reports that affects the local agent itself.

However, I did find an issue where we cannot use --apply on a target without a cache directory at all (never had device run through).

```
Notice: Compiled catalog for cat-6500 in environment production in 0.18 seconds
Debug: Creating default schedules
Error: No such file or directory @ rb_sysopen - /opt/puppetlabs/puppet/cache/devices/cat-6500/state/graphs/resources.dot
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/graph/simple_graph.rb:473:in `initialize'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/graph/simple_graph.rb:473:in `open'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/graph/simple_graph.rb:473:in `write_graph'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/resource/catalog.rb:576:in `write_graph'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/resource/catalog.rb:301:in `finalize'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application/apply.rb:271:in `block in main'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/context.rb:65:in `override'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet.rb:260:in `override'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application/apply.rb:233:in `main'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application/apply.rb:174:in `run_command'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application/device.rb:257:in `block (2 levels) in main'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application/device.rb:228:in `each'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application/device.rb:228:in `collect'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application/device.rb:228:in `block in main'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/context.rb:65:in `override'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet.rb:260:in `override'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application/device.rb:203:in `main'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application.rb:383:in `run_command'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application.rb:375:in `block in run'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util.rb:672:in `exit_on_fail'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/application.rb:375:in `run'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/command_line.rb:137:in `run'
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/util/command_line.rb:73:in `execute'
/opt/puppetlabs/puppet/bin/puppet:5:in `<main>'
```

Reply:
Thanks for looking into that @shermdog. It sounds like the `device` application needs to `Puppet.settings.use` additional sections, but glad to hear state isn't bleeding through to the normal puppet agent/apply case.